lib/rules/no-restricted-imports.js

```diff
@@ -114,38 +132,132 @@ module.exports = {
+                    importNames
+                }
+            });
+        }
+
+        /**
+         * Check if the given importSource is restricted because '*' is being imported.
+         * @param {string} importSource - path of the import
+         * @param {Set.<string>} importNames - Set of import names that are being imported
+         * @returns {boolean} whether the path is restricted
+         * @private
+         */
+        function isRestrictedForEverythingImported(importSource, importNames) {
+            return Object.prototype.hasOwnProperty.call(restrictedPathMessages, importSource) &&
+                restrictedPathMessages[importSource].importNames &&
+                isEverythingImported(importNames);
+        }
+
+        /**
+         * Check if the given importNames are restricted given a list of restrictedImportNames.
+         * @param {Set.<string>} importNames - Set of import names that are being imported
+         * @param {[string]} restrictedImportNames - array of import names that are restricted for this import
+         * @returns {boolean} whether the objectName is restricted
+         * @private
+         */
+        function isRestrictedObject(importNames, restrictedImportNames) {
+            return restrictedImportNames.some(restrictedObjectName => (
+                importNames.has(restrictedObjectName)
+            ));
+        }
+
+        /**
+         * Check if the given importSource is a restricted path.
+         * @param {string} importSource - path of the import
+         * @param {Set.<string>} importNames - Set of import names that are being imported
          * @returns {boolean} whether the variable is a restricted path or not
          * @private
          */
-        function isRestrictedPath(name) {
-            return Object.prototype.hasOwnProperty.call(restrictedPathMessages, name);
+        function isRestrictedPath(importSource, importNames) {
+            let isRestricted = false;
+
+            if (Object.prototype.hasOwnProperty.call(restrictedPathMessages, importSource)) {
+                if (restrictedPathMessages[importSource].importNames) {
+                    isRestricted = isRestrictedObject(importNames, restrictedPathMessages[importSource].importNames);
+                } else {
+                    isRestricted = true;
+                }
+            }
+
+            return isRestricted;
+        }
+
+        /**
+         * Check if the given importSource is restricted by a pattern.
+         * @param {string} importSource - path of the import
+         * @returns {boolean} whether the variable is a restricted pattern or not
+         * @private
+         */
+        function isRestrictedPattern(importSource) {
+            return restrictedPatterns.length > 0 && restrictedPatternsMatcher.ignores(importSource);
         }
 
         return {
             ImportDeclaration(node) {
+
                 if (node && node.source && node.source.value) {
```

Comment:
Are these null checks necessary? I think `node` and `node.source` will always be non-null.

(Actually, it looks like this was in the code before.)

Reply:
Actually, all of the test pass when I completely remove this line, so it seems that an ImportDeclaration `node.source.value` will never be null.  I have made this change with the next commit.